images/virtualization-artifact/pkg/controller/vd/internal/storageclass_ready.go

```diff
@@ -0,0 +1,90 @@
+/*
+Copyright 2024 Flant JSC
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+     http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package internal
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	"github.com/deckhouse/virtualization-controller/pkg/controller/service"
+	virtv2 "github.com/deckhouse/virtualization/api/core/v1alpha2"
+	"github.com/deckhouse/virtualization/api/core/v1alpha2/vdcondition"
+)
+
+type StorageClassReadyHandler struct {
+	service DiskService
+}
+
+func NewStorageClassReadyHandler(diskService DiskService) *StorageClassReadyHandler {
+	return &StorageClassReadyHandler{
+		service: diskService,
+	}
+}
+
+func (h StorageClassReadyHandler) Handle(ctx context.Context, vd *virtv2.VirtualDisk) (reconcile.Result, error) {
+	condition, ok := service.GetCondition(vdcondition.StorageClassReadyType, vd.Status.Conditions)
+	if !ok {
+		condition = metav1.Condition{
+			Type:   vdcondition.StorageClassReadyType,
+			Status: metav1.ConditionUnknown,
+		}
+	}
+
+	defer func() { service.SetCondition(condition, &vd.Status.Conditions) }()
+
+	if vd.DeletionTimestamp != nil {
+		condition.Status = metav1.ConditionUnknown
+		condition.Reason = ""
+		condition.Message = ""
+		return reconcile.Result{}, nil
+	}
+
+	isDefaultStorageClass := vd.Spec.PersistentVolumeClaim.StorageClass == nil || *vd.Spec.PersistentVolumeClaim.StorageClass == ""
+	hasStorageClassInStatus := vd.Status.StorageClassName != ""
+	var checkedSCName *string
+	if hasStorageClassInStatus {
```

Comment:
The condition should reflect the readiness of the StorageClass from the spec, if specified. If the spec has changed, it will reflect the SC from the status, which is incorrect.

Reply:
If spec has changed, CleanUp and Reque after will called